R/new_epipredict_steps/step_yeo_johnson.R

```diff
@@ -0,0 +1,399 @@
+  title <- "Yeo-Johnson transformation (see `lambdas` object for values) on "
+  epipredict:::print_epi_step(x$terms, x$terms, title = title, width = width)
+  invisible(x)
+}
+
+#' Compute the lambda values per group for each column.
+#'
+#' @keywords internal
+#' @rdname recipes-internal
+get_lambdas_yj_table <- function(training, col_names, limits, num_unique, na_lambda_fill, na_rm, epi_keys_checked) {
+  # Estimate the lambda for each column, creating a lambda_ column for each.
+  # Note that estimate_yj() operates on a vector.
+  lambdas <- training %>%
+    summarise(
+      across(all_of(col_names), ~ estimate_yj(.x, limits, num_unique, na_rm)),
+      .by = all_of(epi_keys_checked)
+    ) %>%
+    rename_with(~ paste0(".lambda_", .x), -all_of(epi_keys_checked))
+
+  # Check for NAs in any of the lambda_ columns.
+  # EDIT: This warning was too noisy. Keeping code around, in case we want it.
+  # for (col in col_names) {
+  #   if (any(is.na(values[[paste0("lambda_", col)]]))) {
+  #     cli::cli_warn(
+  #       c(
+  #         x = "Yeo-Johnson lambda could not be estimated for some geos for {col}.",
+  #         i = "Using lambda={x$na_lambda_fill} in these cases."
+  #       ),
+  #       call = rlang::caller_fn()
+  #     )
+  #   }
+  # }
+
+  # Fill in NAs with the default lambda.
+  lambdas %>%
+    mutate(across(starts_with(".lambda_"), \(col) ifelse(is.na(col), na_lambda_fill, col)))
+}
+
+
+### Code below taken from recipes::step_YeoJohnson.
+### https://github.com/tidymodels/recipes/blob/v1.1.1/R/YeoJohnson.R#L172
+
+#' Internal Functions
+#'
+#' @keywords internal
+#' @rdname recipes-internal
+#' @export
+yj_transform <- function(x, lambda, ind_neg = NULL, eps = 0.001) {
+  if (is.na(lambda)) {
+    return(x)
+  }
+  if (!inherits(x, "tbl_df") || is.data.frame(x)) {
+    x <- unlist(x, use.names = FALSE)
+  } else {
+    if (!is.vector(x)) {
+      x <- as.vector(x)
+    }
+  }
+  # TODO case weights: can we use weights here?
+  if (is.null(ind_neg)) {
+    dat_neg <- x < 0
+    ind_neg <- list(is = which(dat_neg), not = which(!dat_neg))
+  }
+  not_neg <- ind_neg[["not"]]
+  is_neg <- ind_neg[["is"]]
+
+  nn_trans <- function(x, lambda) {
+    if (abs(lambda) < eps) {
+      log(x + 1)
+    } else {
+      ((x + 1)^lambda - 1) / lambda
+    }
+  }
+
+  ng_trans <- function(x, lambda) {
+    if (abs(lambda - 2) < eps) {
+      -log(-x + 1)
+    } else {
+      -((-x + 1)^(2 - lambda) - 1) / (2 - lambda)
+    }
+  }
+
+  if (length(not_neg) > 0) {
+    x[not_neg] <- nn_trans(x[not_neg], lambda)
+  }
+
+  if (length(is_neg) > 0) {
+    x[is_neg] <- ng_trans(x[is_neg], lambda)
+  }
+  x
+}
+
+
+## Helper for the log-likelihood calc for eq 3.1 of Yeo, I. K.,
+## & Johnson, R. A. (2000). A new family of power transformations
+## to improve normality or symmetry. Biometrika. page 957
+ll_yj <- function(lambda, y, ind_neg, const, eps = 0.001) {
+  n <- length(y)
+  y_t <- yj_transform(y, lambda, ind_neg)
+  mu_t <- mean(y_t)
```

Comment:
seems like it's inherited, but this is completely unused??

Reply:
yup 🤷

Reply:
ok, so 3.1 involves λ, μ, and v (well, σ^2, but that's just the (v)ariance), but if you substitute things from 3.2 after optimizing over μ and v, the term with μ drops out as a constant, leaving us with 326 (they've already left out n/2*log(2π))